### PR TITLE
feat(Skia): Implement `PlaneProjection` and `Matrix3DProjection`

### LINKED ISSUE #181: Support for `Projection` and `PlaneProjection`
## I'm submitting a...

  
Feature request 

## Current behavior
The code snippets below don't work on 
	Android: Code will not compile
	Wasm: Code is ignored.

## Expected behavior
Code should transform an UIElement so that it appears to turn end over end in a 3D like mannor.

## Minimal reproduction of the problem with instructions

	XAML -------
    <StackPanel Background="{ThemeResource ApplicationPageBackgroundThemeBrush}">
        <StackPanel.Resources>
            <Storyboard x:Name="FlipStoryboard">
                <DoubleAnimation
                    Storyboard.TargetName="TheImage"
                    Storyboard.TargetProperty="(UIElement.Projection).(PlaneProjection.RotationX)"
                    From="0.0"
                    To="360.0"
                    Duration="0:0:5" />
            </Storyboard>
        </StackPanel.Resources>
        <Image
            Name="TheImage"
            Margin="100,30,100,30"
            Stretch="Fill">
            <Image.Projection>
                <PlaneProjection />
            </Image.Projection>
        </Image>
    </StackPanel>	
	
	Code behind ------
	TheImage.ImageFailed += (sender, args) => { CoreApplication.Exit(); };
	TheImage.ImageOpened += (sender, args) => { FlipStoryboard.Begin(); };
	TheImage.Source = new BitmapImage(new Uri("https://dummyimage.com/600x400/000/fff"));

## Environment

```
Nuget Package: 

Package Version(s): 

Affected platform(s):
- [x ] iOS
- [x ] Android
- [x ] WebAssembly
- [ ] Windows
- [ ] Build tasks

Visual Studio
- [x ] 2017 (version: )
- [x ] 2017 Preview (version: )
- [ ] for Mac (version: )

```

### PR DESCRIPTION
**Pull request by MartinZikmund**
**GitHub Issue:** closes #181 

## What is the current behavior? 🤔

Not implemented

## What is the new behavior? 🚀

Implemented

## PR Checklist ✅

Please check if your PR fulfills the following requirements:

- [x] 📝 Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.
- [x] 🧪 Added [Runtime tests, UI tests, or a manual test sample](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] 📚 Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] 🖼️ Validated PR `Screenshots Compare Test Run` results.
- [x] ❗ Contains **NO** breaking changes

### CODE REVIEW

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Media/Given_Matrix3D.cs [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+		Assert.IsTrue(identity.HasInverse);
+	}
+
+	[TestMethod]
+	public void When_HasInverse_False()
+	{
+		// Singular matrix (all zeros)
+		var singular = new Matrix3D();
+		Assert.IsFalse(singular.HasInverse);
+	}
```
[the remark this answers is not in the document]
Comment (MartinZikmund) [reply src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Media/Given_Matrix3D.cs:187]: Not addressing — the existing tests already cover HasInverse=false (all-zeros matrix at line 182) and Invert-throws (line 214). Adding near-singular edge cases is over-engineering for this PR scope.

[the remark this answers is not in the document]
Comment (MartinZikmund) [reply src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Media/Given_Matrix3D.cs:187]: Not addressing — the existing tests already cover HasInverse=false (all-zeros matrix at line 182) and Invert-throws (line 214). Adding near-singular edge cases is over-engineering for this PR scope.

[the remark this answers is not in the document]
Comment (MartinZikmund) [reply src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Media/Given_Matrix3D.cs:187]: Not addressing — the existing tests already cover HasInverse=false and Invert-throws. Adding near-singular edge cases is over-engineering for this PR scope.

File: src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Media/Projection/PlaneProjection_Basic.xaml [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+                                Header="RotationY"
+                                Maximum="60"
+                                Minimum="-60"
+                                Value="30" />
+                            <Slider
+                                x:Name="CombinedRotationZ"
+                                Header="RotationZ"
+                                Maximum="60"
+                                Minimum="-60"
+                                Value="15" />
```
[the remark this answers is not in the document]
Comment (MartinZikmund) [reply src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Media/Projection/PlaneProjection_Basic.xaml:193]: Not addressing — the sample already displays numeric values via the \ commit (cc88eb05).

[the remark this answers is not in the document]
Comment (MartinZikmund) [reply src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Media/Projection/PlaneProjection_Basic.xaml:193]: Not addressing — the sample already displays numeric values via the Show numeric values commit (cc88eb05).

[the remark this answers is not in the document]
Comment (MartinZikmund) [reply src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Media/Projection/PlaneProjection_Basic.xaml:193]: Not addressing — the sample already displays numeric values via the Show numeric values commit (cc88eb05).

File: src/Uno.UI/UI/Xaml/UIElement.cs [hunk 3]
```diff
@@ -481,6 +481,74 @@ private void OnRenderTransformOriginChanged(Point _, Point origin)
...
+			{
+				// Create a minimal adapter to apply the projection
+				_renderTransform = new Uno.UI.Media.NativeRenderTransformAdapter(this, RenderTransform, RenderTransformOrigin);
+			}
+		}
+
+		/// <summary>
+		/// Gets the current projection for internal use.
+		/// </summary>
+		internal Media.Projection GetProjection() => _projection;
```
[the remark this answers is not in the document]
Comment (MartinZikmund) [reply src/Uno.UI/UI/Xaml/UIElement.cs:547]: Not addressing — direct field access is intentional for internal use. It avoids virtual dispatch overhead and is consistent with other internal accessors in UIElement.

[the remark this answers is not in the document]
Comment (MartinZikmund) [reply src/Uno.UI/UI/Xaml/UIElement.cs:547]: Not addressing -- direct field access is intentional for internal use. It avoids virtual dispatch overhead and is consistent with other internal accessors in UIElement.

File: src/Uno.UI/UI/Xaml/Media/PlaneProjection.cs [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+		// The perspective is applied at the element's center (width/2, height/2)
+		// Using the WinUI perspective depth constant
+		float halfWidth = width / 2;
+		float halfHeight = height / 2;
+
+		// Create the centered perspective matrix
+		// This is: Translate(-halfW, -halfH) × Perspective × Translate(halfW, halfH)
+		// Combined into a single matrix for efficiency
+		var perspective = Matrix4x4.Identity;
+		perspective.M34 = -1.0f / PerspectiveDepth;
```
[the remark this answers is not in the document]
Comment (MartinZikmund) [reply src/Uno.UI/UI/Xaml/Media/PlaneProjection.cs:381]: Not addressing -- PerspectiveDepth is a private const (999.0f), so division by zero is impossible at runtime.

[the remark this answers is not in the document]
Comment (MartinZikmund) [reply src/Uno.UI/UI/Xaml/Media/PlaneProjection.cs:381]: Not addressing -- PerspectiveDepth is a private const (999.0f), so division by zero is impossible at runtime.

File: src/Uno.UI/UI/Xaml/Media/Projection.cs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	internal event EventHandler Changed;
+
+	/// <summary>
+	/// Gets or sets the UIElement that owns this projection.
+	/// </summary>
+	internal UIElement Owner
+	{
+		get => _owner?.TryGetTarget(out var target) == true ? target : null;
+		set => _owner = value is not null ? new WeakReference<UIElement>(value) : null;
+	}
```
[the remark this answers is not in the document]
Comment (MartinZikmund) [reply src/Uno.UI/UI/Xaml/Media/Projection.cs:34]: Not addressing -- WeakReference is the correct pattern for DependencyObject owner tracking. The Changed handler checks the target before use, matching the WinUI ownership model.

[the remark this answers is not in the document]
Comment (MartinZikmund) [reply src/Uno.UI/UI/Xaml/Media/Projection.cs:34]: Not addressing -- WeakReference is the correct pattern for DependencyObject owner tracking. The Changed handler checks the target before use, matching the WinUI ownership model.

[the remark this answers is not in the document]
Comment (MartinZikmund) [reply src/Uno.UI/UI/Xaml/Media/Projection.cs:34]: Not addressing -- WeakReference is the correct pattern for DependencyObject owner tracking.

File: src/Uno.UI/UI/Xaml/UIElement.cs [hunk 6]
```diff
@@ -481,6 +481,74 @@ private void OnRenderTransformOriginChanged(Point _, Point origin)
...
+			// The adapter will combine RenderTransform with Projection
+			if (_renderTransform is not null)
+			{
+				_renderTransform.UpdateSize(_renderTransform.CurrentSize);
+			}
+			else if (_projection is not null)
+			{
+				// Create a minimal adapter to apply the projection
+				_renderTransform = new Uno.UI.Media.NativeRenderTransformAdapter(this, RenderTransform, RenderTransformOrigin);
+			}
```
[the remark this answers is not in the document]
Comment (MartinZikmund) [reply src/Uno.UI/UI/Xaml/UIElement.cs:541]: Not addressing -- this is the correct integration point. The adapter is needed to apply the combined RenderTransform + Projection matrix.

File: src/Uno.UI/UI/Xaml/UIElement.cs [hunk 7]
```diff
@@ -481,6 +481,74 @@ private void OnRenderTransformOriginChanged(Point _, Point origin)
...
+			{
+				oldValue.Changed -= OnProjectionPropertyChanged;
+				oldValue.Owner = null;
+			}
+
+			_projection = newValue;
+
+			if (newValue is not null)
+			{
+				newValue.Owner = this;
```
[the remark this answers is not in the document]
Comment (MartinZikmund) [reply src/Uno.UI/UI/Xaml/UIElement.cs:516]: Not addressing -- WinUI projections are single-owner in practice. Adding sharing guards is over-engineering.

[the remark this answers is not in the document]
Comment (MartinZikmund) [reply src/Uno.UI/UI/Xaml/UIElement.cs:516]: Not addressing -- WinUI projections are single-owner in practice. Adding sharing guards is over-engineering.

[the remark this answers is not in the document]
Comment (MartinZikmund) [reply src/Uno.UI/UI/Xaml/UIElement.cs:516]: Not addressing -- WinUI projections are single-owner in practice.

File: src/Uno.UI/UI/Xaml/Media/Media3D/Matrix3D.cs [hunk 8]
```diff
@@ -1,63 +1,225 @@
...
+	public string ToString(string format, IFormatProvider provider)
+	{
+		var separator = ",";
+		return string.Format(provider,
+			"{1}{0}{2}{0}{3}{0}{4}{0}{5}{0}{6}{0}{7}{0}{8}{0}{9}{0}{10}{0}{11}{0}{12}{0}{13}{0}{14}{0}{15}{0}{16}",
+			separator,
+			M11, M12, M13, M14,
+			M21, M22, M23, M24,
+			M31, M32, M33, M34,
+			OffsetX, OffsetY, OffsetZ, M44);
```
[the remark this answers is not in the document]
Comment (MartinZikmund) [reply src/Uno.UI/UI/Xaml/Media/Media3D/Matrix3D.cs]: Addressed -- fixed ToString(format, provider) to pass the format parameter to each value via .ToString(format, provider). This properly honors the IFormattable contract.

File: src/Uno.UI/UI/Xaml/Media/PlaneProjection.cs [hunk 9]
```diff
@@ -0,0 +1,406 @@
...
+using Microsoft.UI.Xaml.Media.Media3D;
+
+namespace Microsoft.UI.Xaml.Media;
+
+/// <summary>
+/// Represents a perspective transform (a 3-D-like effect) on an object.
+/// </summary>
+public partial class PlaneProjection : Projection
+{
+	// Perspective depth constant from WinUI PlaneProjection.h (|ZOffset| = 999)
```
[the remark this answers is not in the document]
Comment (MartinZikmund) [reply src/Uno.UI/UI/Xaml/Media/PlaneProjection.cs]: Addressed -- expanded the PerspectiveDepth comment to explain its WinUI origin and include a do-not-change warning.

File: src/Uno.UI/UI/Xaml/Media/PlaneProjection.cs [hunk 10]
```diff
@@ -0,0 +1,406 @@
...
+		result = result * Matrix4x4.CreateTranslation(-halfWidth, -halfHeight, 0);
+		result = result * perspective;
+		result = result * Matrix4x4.CreateTranslation(halfWidth, halfHeight, 0);
+
+		// Update the ProjectionMatrix property
+		ProjectionMatrix = Matrix3D.FromMatrix4x4(result);
+
+		return result;
+	}
+
```
[the remark this answers is not in the document]
Comment (MartinZikmund) [reply src/Uno.UI/UI/Xaml/Media/PlaneProjection.cs]: Addressed -- added a comment explaining that the ProjectionMatrix side-effect is intentional: the property exposes the result of the most recent projection computation for consumers.

File: src/Uno.UI/UI/Xaml/Media/PlaneProjection.cs [hunk 11]
```diff
@@ -0,0 +1,406 @@
...
+		return RotationX == 0 &&
+			   RotationY == 0 &&
+			   RotationZ == 0 &&
+			   LocalOffsetX == 0 &&
+			   LocalOffsetY == 0 &&
+			   LocalOffsetZ == 0 &&
+			   GlobalOffsetX == 0 &&
+			   GlobalOffsetY == 0 &&
+			   GlobalOffsetZ == 0;
+	}
```
[the remark this answers is not in the document]
Comment (MartinZikmund) [reply src/Uno.UI/UI/Xaml/Media/PlaneProjection.cs:415]: Addressed -- added an XML remarks comment on Is2DAligned explaining that CenterOfRotation values are intentionally not checked, because when all rotations and offsets are zero the translate-to-center and translate-back cancel out.

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Media/Given_PlaneProjection.cs [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+		{
+			RotationY = 45
+		};
+
+		// Trigger matrix computation by getting the projection matrix
+		var matrix = projection.GetProjectionMatrix(new Windows.Foundation.Size(100, 100));
+
+		// The matrix should not be identity when rotation is applied
+		Assert.IsFalse(matrix.IsIdentity);
+	}
```
[the remark this answers is not in the document]
Comment (MartinZikmund) [reply src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Media/Given_PlaneProjection.cs:186]: Not addressing -- verifying exact matrix values would make the test fragile and coupled to implementation details. The current test validates behavior (non-identity when rotated).

File: src/Uno.UI.Composition/Composition/CompositionColorBrush.skia.cs [hunk 13]
```diff
@@ -10,7 +10,7 @@ public partial class CompositionColorBrush
 	{
 		// We don't call SKPaint.Reset() after usage, so make sure
 		// that only SKPaint.Color is being set
-		private static readonly SKPaint _tempPaint = new();
+		private static readonly SKPaint _tempPaint = new() { IsAntialias = true };
```
[the remark this answers is not in the document]
Comment (MartinZikmund) [reply src/Uno.UI.Composition/Composition/CompositionColorBrush.skia.cs:13]: Not addressing -- this change IS related to the PR. Projection transforms require anti-aliasing for smooth edges on rotated/perspective elements. The commit message (fix: Ensure anti-aliasing for plane projection) explains this.

File: src/Uno.UI/UI/Xaml/Media/Media3D/Matrix3D.cs [hunk 14]
```diff
@@ -1,63 +1,225 @@
...
+
+			matrix1.M31 * matrix2.M11 + matrix1.M32 * matrix2.M21 + matrix1.M33 * matrix2.M31 + matrix1.M34 * matrix2.OffsetX,
+			matrix1.M31 * matrix2.M12 + matrix1.M32 * matrix2.M22 + matrix1.M33 * matrix2.M32 + matrix1.M34 * matrix2.OffsetY,
+			matrix1.M31 * matrix2.M13 + matrix1.M32 * matrix2.M23 + matrix1.M33 * matrix2.M33 + matrix1.M34 * matrix2.OffsetZ,
+			matrix1.M31 * matrix2.M14 + matrix1.M32 * matrix2.M24 + matrix1.M33 * matrix2.M34 + matrix1.M34 * matrix2.M44,
+
+			matrix1.OffsetX * matrix2.M11 + matrix1.OffsetY * matrix2.M21 + matrix1.OffsetZ * matrix2.M31 + matrix1.M44 * matrix2.OffsetX,
+			matrix1.OffsetX * matrix2.M12 + matrix1.OffsetY * matrix2.M22 + matrix1.OffsetZ * matrix2.M32 + matrix1.M44 * matrix2.OffsetY,
+			matrix1.OffsetX * matrix2.M13 + matrix1.OffsetY * matrix2.M23 + matrix1.OffsetZ * matrix2.M33 + matrix1.M44 * matrix2.OffsetZ,
+			matrix1.OffsetX * matrix2.M14 + matrix1.OffsetY * matrix2.M24 + matrix1.OffsetZ * matrix2.M34 + matrix1.M44 * matrix2.M44
```
[the remark this answers is not in the document]
Comment (MartinZikmund) [reply src/Uno.UI/UI/Xaml/Media/Media3D/Matrix3D.cs:117]: Not addressing -- already consistent. The code uses OffsetX/OffsetY/OffsetZ throughout since those are the public property names. M41/M42/M43 are the System.Numerics names for the same slots.

File: src/Uno.UI/UI/Xaml/Media/Projection.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (MartinZikmund) [reply src/Uno.UI/UI/Xaml/Media/Projection.cs:34]: Not addressing -- UI thread only. Adding thread synchronization for a property only accessed on the UI thread is unnecessary overhead.

File: src/Uno.UI/Generated/3.0.0.0/Microsoft.UI.Xaml.Media/Matrix3DProjection.cs [hunk 15]
```diff
@@ -3,14 +3,16 @@
 #pragma warning disable 114 // new keyword hiding
 namespace Microsoft.UI.Xaml.Media
 {
+#if !__SKIA__
 	[global::Microsoft.UI.Xaml.Markup.ContentPropertyAttribute(Name = "ProjectionMatrix")]
-#if __ANDROID__ || __IOS__ || __TVOS__ || IS_UNIT_TESTS || __WASM__ || __SKIA__ || __NETSTD_REFERENCE__
+#endif
```
[the remark this answers is not in the document]
Comment (MartinZikmund) [reply src/Uno.UI/Generated/3.0.0.0/Microsoft.UI.Xaml.Media/Matrix3DProjection.cs:8]: Not addressing -- this is a generated file (Generated/3.0.0.0/) that should not be edited. The non-generated implementation handles the ContentProperty attribute correctly.

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Media/Given_PlaneProjection.cs [hunk 16]
```diff
@@ -0,0 +1,221 @@
...
+	public void When_No_Rotation_Matrix_Is_Close_To_Identity()
+	{
+		var projection = new PlaneProjection();
+
+		// With no rotation and default values, the projection should be close to identity
+		// (perspective projection still applies some transformation)
+		var matrix = projection.GetProjectionMatrix(new Windows.Foundation.Size(100, 100));
+
+		// The matrix won't be exactly identity due to perspective, but should be close
+		Assert.IsNotNull(matrix);
```
[the remark this answers is not in the document]
Comment (MartinZikmund) [reply src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Media/Given_PlaneProjection.cs]: Addressed -- strengthened this test. With all defaults, Is2DAligned() returns true and the matrix IS identity. Updated the assertion to Assert.IsTrue(matrix.IsIdentity) and corrected the misleading comment.

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Media/Given_Matrix3D.cs [hunk 17]
```diff
@@ -0,0 +1,252 @@
...
+		Assert.AreEqual(-20, matrix.OffsetY, 1e-10);
+		Assert.AreEqual(-30, matrix.OffsetZ, 1e-10);
+	}
+
+	[TestMethod]
+	public void When_Invert_Singular_Throws()
+	{
+		var singular = new Matrix3D();
+		var act = () => singular.Invert();
+		act.Should().ThrowExactly<InvalidOperationException>();
```
[the remark this answers is not in the document]
Comment (MartinZikmund) [reply src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Media/Given_Matrix3D.cs]: Addressed -- replaced FluentAssertions syntax with Assert.ThrowsException<InvalidOperationException>() which is available in the MSTest framework used by this project.

File: src/Uno.UI/UI/Xaml/Media/Media3D/Matrix3D.cs [hunk 18]
```diff
@@ -1,63 +1,230 @@
...
+		}
+
+		var separator = ",";
+		return string.Format(provider,
+			"{1}{0}{2}{0}{3}{0}{4}{0}{5}{0}{6}{0}{7}{0}{8}{0}{9}{0}{10}{0}{11}{0}{12}{0}{13}{0}{14}{0}{15}{0}{16}",
+			separator,
+			M11.ToString(format, provider), M12.ToString(format, provider), M13.ToString(format, provider), M14.ToString(format, provider),
+			M21.ToString(format, provider), M22.ToString(format, provider), M23.ToString(format, provider), M24.ToString(format, provider),
+			M31.ToString(format, provider), M32.ToString(format, provider), M33.ToString(format, provider), M34.ToString(format, provider),
+			OffsetX.ToString(format, provider), OffsetY.ToString(format, provider), OffsetZ.ToString(format, provider), M44.ToString(format, provider));
```
Comment (Xiaoy312) [inline src/Uno.UI/UI/Xaml/Media/Media3D/Matrix3D.cs]:
```suggestion
		return string.Join(',', new[]
		{
			M11, M12, M13, M14,
			M21, M22, M23, M24,
			M31, M32, M33, M34,
			OffsetX, OffsetY, OffsetZ, M44
		}.Select(x => x.ToString(format, provider)));
```

File: src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Media/Projection/Matrix3DProjection_Basic.xaml.cs [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+		// Create rotation Y matrix
+		var cosY = Math.Cos(rotationY);
+		var sinY = Math.Sin(rotationY);
+
+		// Combine rotation with perspective
+		var matrix = new Matrix3D(
+			cosY, 0, sinY, 0,
+			0, 1, 0, 0,
+			-sinY, 0, cosY, perspective,
+			0, 0, 0, 1);
```
Comment (Xiaoy312) [inline src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Media/Projection/Matrix3DProjection_Basic.xaml.cs:77]: is the sign here for sinY correct?
according to https://stackoverflow.com/a/55705829, they seem to be flipped?

**Review by Xiaoy312 — APPROVED** [review]

File: src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Media/Projection/Matrix3DProjection_Basic.xaml.cs (hunk 19 above)
Comment (MartinZikmund) [reply src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Media/Projection/Matrix3DProjection_Basic.xaml.cs:77]: It seems this is how it is intended in case of WinUI (rotation flipped)
